### ISSUE TITLE
Subgraph Execution

### ISSUE THREAD

**Issue by ShyamW** [ARM-software/armnn#223]
I was wondering if Armnn had the ability to execute a tensorflow subgraph. In other words, the input layer wouldn't be named input and could be the 2nd, 3rd, ... , layer in the graph

**Comment by MatthewARM** [comment]
That might work out of the box, try passing the name of the layer to the parser. Otherwise I usually use the tensorflow graph_transform tool to produce the cut-down graph: https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms#using-the-graph-transform-tool

This lets you set exactly which nodes you want as the inputs and outputs of the subgraph.

**Comment by ShyamW** [comment]
When I tried out of the box (from the full graph), I get this error:

`terminate called after throwing an instance of 'armnn::ParseException'
  what():  Missing input shape for Placeholder 'input'  at function ParsePlaceholder [/oe/bld/build-CORTEX_1/arago-tmp-external-linaro-toolchain/work/armv7ahf-neon-linux-gnueabi/armnn/19.02-r0/git/src/armnnTfParser/TfParser.cpp:2513]`


When I tried to cut the graph using tensorflow's optimize_for_inference.py:
`python3 optimize_for_inference.py --input=mobilenet_v1_1.0_224_frozen.pb --output=arm_sub2.pb --input_names=MobilenetV1/Logits/AvgPool_1a/AvgPool --output_names=MobilenetV1/Predictions/Softmax --frozone_graph=True`

The graph would cut correctly, but when I tried to pass data to the AvgPool Layer, I got this error:
`terminate called after throwing an instance of 'armnn::Exception'
  what():  Optimize returned nullptr
`

**Comment by ShyamW** [comment]
@MatthewARM 
I tried to use this command:


`bazel-bin/tensorflow/tools/graph_transforms/transform_graph --in_graph=tensorflow/python/tools/mobilenet_v1_1.0_224_frozen.pb --out_graph=~/transf.pb --inputs='MobilenetV1/Logits/AvgPool_1a/AvgPool' --outputs='MobilenetV1/Predictions/Softmax' --transforms='
  sort_by_execution_order'`


However, I am still getting this error which leads me to believe it did not cut the graph (since the error is the same as the "out of the box method":

`terminate called after throwing an instance of 'armnn::ParseException'
  what():  Missing input shape for Placeholder 'input'  at function ParsePlaceholder [/oe/bld/build-CORTEX_1/arago-tmp-external-linaro-toolchain/work/armv7ahf-neon-linux-gnueabi/armnn/19.02-r0/git/src/armnnTfParser/TfParser.cpp:2513]`


Any help is greatly appreciated!

**Comment by MatthewARM** [comment]
That's a bit odd, "terminate called after throwing an instance of 'armnn::Exception' what(): Optimize returned nullptr" isn't a very helpful error message. Can you turn on more verbose logging to see if there's anything more helpful first? (See include/armnn/Utils.hpp, probably setting LogSeverity::Debug would be enough).

**Comment by ShyamW** [comment]
I added this line to my code:
`
armnn::LogSeverity level = armnn::LogSeverity::Debug;
`

However, I am still getting the same error

**Comment by MatthewARM** [comment]
Hi @ShyamW it should look something like this:

```
armnn::ConfigureLogging(true, false, armnn::LogSeverity::Debug);
```

That should make it produce debugging information on the standard output.

**Comment by ShyamW** [comment]
Ah thanks, adding that produced this error:

WARNING: Layer of type Convolution2d is not supported on requested backend CpuAcc for data type Float32 (reason: in validate src/runtime/NEON/functions/NEGEMMConvolutionLayer.cpp:444: weights->dimension(idx_channel) != input->dimension(idx_channel)), falling back to the next backend.
ERROR: Layer of type Convolution2d is not supported on any preferred backend [CpuAcc ]

**Comment by MatthewARM** [comment]
Hi @ShyamW ok that's good information, now we know that for some reason the Compute Library is rejecting one of the convolution layers. 

Looking at your layers list, I can see you have a lot of unknown tensor dimensions. I guess that this could be the problem as it is unsupported by Arm NN at this time (I'm hoping we can add support for this over the next few months). Is it necessary for your use case to have dynamic tensor dimensions?  If not, you can fix the dimensions of the input tensor by passing ```--transforms='strip_unused_nodes(type=float, shape="<new size here>")``` to graph_transform. There's an example at https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/graph_transforms/README.md#optimizing-for-deployment

**Comment by MatthewARM** [comment]
@ShyamW I hope yoy are well. I'm closing this tickeg as we have no plan at this time to support execution of tensorflow subgraphs via armnn. This model also needs dynamic tensors, which are under development.
